presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PhysicalResourceOptimizer.java

```diff
@@ -0,0 +1,308 @@
+import static com.facebook.presto.SystemSessionProperties.isResourceManagementComputationEnabled;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class PhysicalResourceOptimizer
+        implements PlanOptimizer
+{
+    private static final Logger log = Logger.get(PhysicalResourceOptimizer.class);
+    private static final int PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_LARGER_QUERIES = 4096;
+    private static final int PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_LARGER_QUERIES = 4096;
+    private static final int PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_LARGER_QUERIES = 1024;
+    private static final int PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_LARGER_QUERIES = 4096;
+    private static final int PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_SHORTER_QUERIES = 1024;
+    private static final int PARTITION_COUNT_FOR_DIRCU_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_SHORTER_QUERIES = 4096;
+    private static final int PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MAX_EXECUTOR_FOR_SHORTER_QUERIES = 4096;
+    private static final int PARTITION_COUNT_FOR_LATENCY_OPTIMIZED_FOR_MIN_EXECUTOR_FOR_SHORTER_QUERIES = 4096;
+    // This needs to be extracted from  "mode": "T1_MIGRATION_10G",
+    private static final double CONTAINER_SIZE = 10.0;
+    private static final double THRESHOLD_SIZE_FOR_LONGER_QUERIES = 600; // in GB
+    private static final int MAX_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED = 600;
+    private static final int MIN_EXECUTOR_FOR_LARGE_QUERIES_DIRCU_OPTIMIZED = 200;
+    private static final int MAX_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED = 1000;
+    private static final int MIN_EXECUTOR_FOR_LARGE_QUERIES_LATENCY_OPTIMIZED = 400;
+    private static final int MAX_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED = 200;
+    private static final int MIN_EXECUTOR_FOR_SMALL_QUERIES_DIRCU_OPTIMIZED = 50;
+    private static final int MAX_EXECUTOR_FOR_SMALL_QUERIES_LATENCY_OPTIMIZED = 400;
+    private static final int MIN_EXECUTOR_FOR_SMALL_QUERIES_LATENCY_OPTIMIZED = 200;
+
+    private final Metadata metadata;
+    private int executorCount;
+    private int hashPartitionCount;
+
+    public PhysicalResourceOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.executorCount = 0;
+        this.hashPartitionCount = 0;
+    }
+
+    public int getExecutorCount()
+    {
+        return this.executorCount;
+    }
+
+    public int getHashPartitionCount()
+    {
+        return this.hashPartitionCount;
+    }
+
+    public boolean isValid()
+    {
+        return ((this.executorCount > 0) && (this.hashPartitionCount > 0));
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, PlanVariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
```

Comment:
The optimize needed to be overridden because of the PlanOptimizer. Now that we have got ridden of the PlanOptimizer interface we have a calculate function returning  `PhysicalResourceSettings`